### PR TITLE
Add support for Protobuf and JSON Schema

### PR DESCRIPTION
**Pull request by rayokota**

### CODE REVIEW

**Comment by rigelbm** [comment]
Hi @rayokota. Sorry for the delay, this is a pretty big PR. I have one big and easy comment to make before any other: could you please not make any changes to the V1 API? That means reverting all changes to kafka-rest-scala-consumers, and the relevant HTTP endpoints. V1 API is currently deprecated, and we plan to remove it **very** soon. That will make review much easier, as it removes half of the changes. Thanks! :smile:

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/AvroConverter.java [hunk 1]
```diff
@@ -18,76 +18,40 @@
...
     try {
-      ByteArrayOutputStream out = new ByteArrayOutputStream();
-      jsonMapper.writeValue(out, value);
-      DatumReader<Object> reader = new GenericDatumReader<Object>(schema);
-      Object object = reader.read(
-          null, decoderFactory.jsonDecoder(schema, new ByteArrayInputStream(out.toByteArray())));
-      out.close();
-      return object;
-    } catch (IOException | RuntimeException e) {
+      return AvroSchemaUtils.toObject(value, (AvroSchema) parsedSchema);
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/AvroConverter.java]: Either rename AvroSchemaUtils to AvroUtils, and ProtobufSchemaUtils to ProtobufUtils, or rename JsonSchemaUtils to JsonSchemaSchemaUtils.

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/AvroConverter.java [hunk 2]
```diff
@@ -18,76 +18,40 @@
...
 import java.io.IOException;
-import java.nio.ByteBuffer;
-import java.util.HashMap;
-import java.util.Map;
 
 /**
  * Provides conversion of JSON to/from Avro.
  */
-public class AvroConverter {
+public class AvroConverter implements SchemaConverter {
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/AvroConverter.java]: Make class final.

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/JsonSchemaConverter.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import io.confluent.kafka.schemaregistry.json.JsonSchemaUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Provides conversion of JSON to/from JSON Schema.
+ */
+public class JsonSchemaConverter implements SchemaConverter {
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/JsonSchemaConverter.java]: Make class final.

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/JsonSchemaConverter.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+  private static final Logger log = LoggerFactory.getLogger(JsonSchemaConverter.class);
+
+  private static final ObjectMapper jsonMapper = new ObjectMapper();
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/JsonSchemaConverter.java]: Rename jsonMapper to JSON_MAPPER.

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/ProtobufConverter.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import io.confluent.kafka.schemaregistry.protobuf.ProtobufSchemaUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Provides conversion of JSON to/from Protobuf.
+ */
+public class ProtobufConverter implements SchemaConverter {
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/ProtobufConverter.java]: Make class final.

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/ProtobufConverter.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+  private static final Logger log = LoggerFactory.getLogger(ProtobufConverter.class);
+
+  private static final ObjectMapper jsonMapper = new ObjectMapper();
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/ProtobufConverter.java]: Rename jsonMapper to JSON_MAPPER.

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/AvroConverter.java [hunk 7]
```diff
@@ -18,76 +18,40 @@
...
 
   private static final Logger log = LoggerFactory.getLogger(AvroConverter.class);
 
-  private static final EncoderFactory encoderFactory = EncoderFactory.get();
-  private static final DecoderFactory decoderFactory = DecoderFactory.get();
   private static final ObjectMapper jsonMapper = new ObjectMapper();
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/AvroConverter.java]: Rename jsonMapper to JSON_MAPPER.

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/JsonSchemaConverter.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+      byte[] bytes = JsonSchemaUtils.toJson(value);
+      if (bytes == null) {
+        return new JsonNodeAndSize(null, 0);
+      }
+      return new JsonNodeAndSize(jsonMapper.readTree(bytes), bytes.length);
+    } catch (IOException e) {
+      log.error("Jackson failed to deserialize JSON: ", e);
+      throw new ConversionException("Failed to convert JSON Schema to JSON: " + e.getMessage());
+    } catch (RuntimeException e) {
+      log.error("Unexpected exception convertion JSON Schema to JSON: ", e);
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/JsonSchemaConverter.java]: s/convertion/converting/

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/JsonSchemaConverter.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      throw new ConversionException("Failed to convert JSON Schem to JSON: " + e.getMessage());
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/JsonSchemaConverter.java]: s/Schem/Schema/

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/JsonSchemaConverter.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2018 Confluent Inc.
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/JsonSchemaConverter.java]: s/2018/2020/

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/ProtobufConverter.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2018 Confluent Inc.
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/ProtobufConverter.java]: s/2018/2020/

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/ProtobufConverter.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+      byte[] bytes = ProtobufSchemaUtils.toJson((Message) value);
+      if (bytes == null) {
+        return new JsonNodeAndSize(null, 0);
+      }
+      return new JsonNodeAndSize(jsonMapper.readTree(bytes), bytes.length);
+    } catch (IOException e) {
+      log.error("Jackson failed to deserialize JSON generated by Protobuf's JSON encoder: ", e);
+      throw new ConversionException("Failed to convert Protobuf to JSON: " + e.getMessage());
+    } catch (RuntimeException e) {
+      log.error("Unexpected exception convertion Protobuf to JSON: ", e);
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/ProtobufConverter.java]: s/convertion/converting/

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/SchemaConverter.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import io.confluent.kafka.schemaregistry.ParsedSchema;
+
+/**
+ * Provides conversion of JSON to/from an object for a specific schema type.
+ */
+public interface SchemaConverter {
+
+  Object toObject(JsonNode value, ParsedSchema schema);
+
+  class JsonNodeAndSize {
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/SchemaConverter.java]: Make class final.

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/SchemaConverter.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+    public JsonNode json;
+    public long size;
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/SchemaConverter.java]: Make json and size attributes private and final and add getters.

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/SchemaConverter.java (hunk 13 above)
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/SchemaConverter.java]: Move class declaration to the end of SchemaConverter.

File: kafka-rest/src/main/java/io/confluent/kafkarest/resources/TopicsResource.java [hunk 15]
```diff
@@ -136,6 +136,60 @@ public void produceAvro(
     produce(asyncResponse, topicName, EmbeddedFormat.AVRO, request);
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/resources/TopicsResource.java]: All three methods have basically the same implementation. Can you Extract-Method?

File: kafka-rest/src/main/java/io/confluent/kafkarest/resources/v2/PartitionsResource.java [hunk 16]
```diff
@@ -138,6 +141,62 @@ public void produceAvro(
     produce(asyncResponse, topic, partition, EmbeddedFormat.AVRO, request);
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/resources/v2/PartitionsResource.java]: Ditto about Extract-Method.

File: kafka-rest/src/main/java/io/confluent/kafkarest/v2/SchemaKafkaConsumerState.java [hunk 17]
```diff
@@ -19,29 +19,35 @@
...
- * Avro implementation of KafkaConsumerState, which decodes into GenericRecords or primitive types.
+ * Schema-specific implementation of KafkaConsumerState, which decodes
+ * into Objects or primitive types.
  */
-public class AvroKafkaConsumerState extends KafkaConsumerState<Object, Object, JsonNode, JsonNode> {
+public class SchemaKafkaConsumerState
+    extends KafkaConsumerState<Object, Object, JsonNode, JsonNode> {
 
-  public AvroKafkaConsumerState(KafkaRestConfig config,
+  private SchemaConverter schemaConverter;
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/v2/SchemaKafkaConsumerState.java]: Make schemaConverter attribute final.

File: kafka-rest/src/main/java/io/confluent/kafkarest/v2/SchemaKafkaConsumerState.java [hunk 18]
```diff
@@ -19,29 +19,35 @@
...
 import org.apache.kafka.clients.consumer.Consumer;
 import org.apache.kafka.clients.consumer.ConsumerRecord;
 
 /**
- * Avro implementation of KafkaConsumerState, which decodes into GenericRecords or primitive types.
+ * Schema-specific implementation of KafkaConsumerState, which decodes
+ * into Objects or primitive types.
  */
-public class AvroKafkaConsumerState extends KafkaConsumerState<Object, Object, JsonNode, JsonNode> {
+public class SchemaKafkaConsumerState
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/v2/SchemaKafkaConsumerState.java]: Make class final.

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/AvroConverter.java (hunk 1 above)
Comment (rayokota) [reply kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/AvroConverter.java]: These classes are in Schema Registry and match the `AvroSchema`, `ProtobufSchema`, and `JsonSchema` classes.  I think I'll leave the current naming scheme for now.

File: kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/SchemaConverter.java [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+  /**
+   * Converts data (including primitive types) to their equivalent JsonNode representation.
+   *
+   * @param value the value to convert
+   * @return an object containing the root JsonNode representing the converted object and the size
+   *     in bytes of the data when serialized
+   */
+  JsonNodeAndSize toJson(Object value);
+
+  final class JsonNodeAndSize {
```
Comment (rigelbm) [inline kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/SchemaConverter.java]: Do you really need this class to be mutable? I don't see the setters being used anywhere. Please, make attributes final and remove setters.

Comment (rayokota) [reply kafka-rest-common/src/main/java/io/confluent/kafkarest/converters/SchemaConverter.java]: Ack

File: kafka-rest/src/test/java/io/confluent/kafkarest/unit/JsonSchemaConverterTest.java [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+import static org.junit.Assert.fail;
+
+public class JsonSchemaConverterTest {
+
+  private static final String recordSchemaString =
+      "{\"properties\": {\n"
+      + "     \"null\": {\"type\": \"null\"},\n"
+      + "     \"boolean\": {\"type\": \"boolean\"},\n"
+      + "     \"number\": {\"type\": \"number\"},\n"
+      + "     \"string\": {\"type\": \"string\"},\n"
```
Comment (rigelbm) [inline kafka-rest/src/test/java/io/confluent/kafkarest/unit/JsonSchemaConverterTest.java]: FYI: JsonSchemaConverterTest does not work on my machine (even though it's passing on Jenkins!!). There's an extra comma in the end of this line that makes this an invalid JSON. The whole test fails to initialise.

Comment (rayokota) [reply kafka-rest/src/test/java/io/confluent/kafkarest/unit/JsonSchemaConverterTest.java]: Ack.  I just switched to a stricter JSON parser, that's why

**Pull request merged**